Researcher profile REST repository

LucaGiamminonni opened the pull request:
## References
Related to https://github.com/DSpace/DSpace/issues/8157
* Related to Contract: https://github.com/DSpace/RestContract/pull/180
* Requires UI PR: https://github.com/DSpace/dspace-angular/pull/1617

## Description
This PR introduces:

- the researcher profile repository
- two features (CanClaimItemFeature and ShowClaimItemFeature) relating to the claim operation of a specific item (generally of the Person type)

## Enabling this feature
Researcher Profiles are _disabled by default._ To enable, uncomment this setting (or add to your local.cfg):
```
researcher-profile.entity-type = Person
```
You also MUST enable the EPersonAuthority plugin & configure the "dspace.object.owner" field.
```
plugin.named.org.dspace.content.authority.ChoiceAuthority = \
org.dspace.content.authority.EPersonAuthority = EPersonAuthority

# Configuration settings required for Researcher Profiles
# These settings ensure "dspace.object.owner" field are indexed by Authority Control
choices.plugin.dspace.object.owner = EPersonAuthority
choices.presentation.dspace.object.owner = suggest
authority.controlled.dspace.object.owner = true
```
NOTE: After updating this value you must restart Tomcat!

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

Commit:
[CST-5306] Migrate Researcher Profile (REST) .

Commit:
[CST-5306] Migrate Researcher Profile (REST)

eskander17 committed:
[CST-5306] Migrate Researcher Profile (REST).
added ClaimItemFeature.

eskander17 committed:
[CST-5306] Migrate Researcher Profile (REST).
- added new four test functions for claim profile.
- formatting the code.
- removed unused classes.

Commit:
Merge remote-tracking branch '4science-bitbucket/main' into CST-5306

Commit:
[CST-5306] Minor improvements in ResearcherProfileServiceImpl

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResearcherProfileRestRepositoryIT.java:
```diff
@@ -0,0 +1,1049 @@
+
+        String adminToken = getAuthToken(admin.getEmail(), password);
+
+        // create and delete a profile
```
This process of first creating & deleting a profile using the REST API is unrealistic for this scenario, as it's very unlikely for a profile to be autoclaimed in this way.  It's much more likely that an Entity of type Person is created that happens to have a matching email address.   

So, I'd recommend we change this to create an Entity of type Person (using the Entity endpoints) and add the user's email address to it.  Then we can prove that Entity will be auto-claimed on next login.

If you'd rather keep this method as-is, then I'd recommend we have a different method test the process I'm describing.  I think we need a test that proves that a regular Entity can be turned into / auto-claimed as a Profile

LucaGiamminonni replied:
I will create a new test

tdonohue commented on dspace/config/modules/authority.cfg:
```diff
@@ -0,0 +1,15 @@
+#####  Authority Control Settings  #####
+
+plugin.named.org.dspace.content.authority.ChoiceAuthority = \
+org.dspace.content.authority.EPersonAuthority = EPersonAuthority
```
While I'm not against creating a separate `authority.cfg` file... I wanted to note that we already have other `ChoiceAuthority` example settings in `dspace.cfg`.  

So, we may want to move the other `ChoiceAuthority` settings from `dspace.cfg` over to this file as well.  Either that, or move these settings to `dspace.cfg`.  It doesn't make sense to have some Authority settings in dspace.cfg while others are in authority.cfg.

Here's the authority control settings in dspace.cfg: https://github.com/DSpace/DSpace/blob/main/dspace/config/dspace.cfg#L1379-L1465

LucaGiamminonni replied:
I will move the already existing example in the new authority.cfg

tdonohue commented on dspace/config/modules/authority.cfg:
```diff
@@ -0,0 +1,15 @@
+
+choices.plugin.dspace.object.owner = EPersonAuthority
+choices.presentation.dspace.object.owner = suggest
+authority.controlled.dspace.object.owner = true
```
These 3 settings have no description. They need a description so that others know how they are used & what values they may accept.  They are not self explanatory

tdonohue replied:
These three new settings still need a description and examples of possible values.  They are undocumented currently.

tdonohue replied:
@LucaGiamminonni : I think you misunderstand my point.  It appears these three settings are **required** for  Researcher Profiles to work properly.  If you notice, in every other example in this "authority.cfg", there's some sort of basic comment that says things like..

```
# Configuration settings for ORCID based authority control.
# Uncomment the lines below to enable configuration
```

But, in this code, you've given users no information about these settings.  I think they MUST have comments to describe their purpose. 

Please add something like:
```
# Configuration settings required for Researcher Profiles
# These settings ensure "dspace.object.owner" field are indexed by Authority Control
```
(I'm not sure if that text is correct, but that's essentially what I'm asking for.  These settings need to document how/why they are used. That's required of all configurations.)

tdonohue reviewed (changes requested):
@LucaGiamminonni : Overall, good work here!  I gave this a code review today though and came up with a lot of questions/comments inline.  It's possible some of these are misunderstandings of the code, but, if so, then I think we need clearer JavaDocs / code comments to describe what the behavior is *supposed to be*.  I found JavaDocs were lacking in some places or were not very detailed...so, it forced me to have to read the code to determine what I *thought it might be doing*.

There's a lot of inline comments below (over 30), but I think most are relatively small.  Glad to chat more though if I misunderstood or my comment doesn't make sense.

Commit:
Merge branch 'main' into CST-5306

Commit:
Merge remote-tracking branch '4science-bitbucket/main' into CST-5306

LucaGiamminonni commented on dspace-api/src/main/java/org/dspace/app/profile/ResearcherProfileServiceImpl.java:
```diff
@@ -0,0 +1,293 @@
+        WorkspaceItem workspaceItem = workspaceItemService.create(context, collection, true);
+        Item item = workspaceItem.getItem();
+        itemService.addMetadata(context, item, "dc", "title", null, null, fullName);
+        itemService.addMetadata(context, item, "dspace", "object", "owner", null, fullName, id, CF_ACCEPTED);
```
Because the entity type is inherited from the collection

LucaGiamminonni commented on dspace-api/src/main/java/org/dspace/app/profile/ResearcherProfileServiceImpl.java:
```diff
@@ -0,0 +1,293 @@
+
+        Group anonymous = groupService.findByName(context, ANONYMOUS);
+        authorizeService.removeGroupPolicies(context, item, anonymous);
+        authorizeService.addPolicy(context, item, READ, ePerson);
```
By default the profile is private, I can add a configuration for this.

LucaGiamminonni commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResearcherProfileRestRepositoryIT.java:
```diff
@@ -0,0 +1,1049 @@
+
+        String authToken = getAuthToken(user.getEmail(), password);
+
+        getClient(authToken).perform(post("/api/eperson/profiles/")
```
The collection that contains the profiles is deleted using the builder and in cascade all the items inside it are also deleted (see CollectionServiceImpl.delete)

LucaGiamminonni commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/ShowClaimItemFeature.java:
```diff
@@ -0,0 +1,93 @@
+@Component
+@AuthorizationFeatureDocumentation(name = ShowClaimItemFeature.NAME,
+    description = "Used to verify if the given user can request the claim of an item")
+public class ShowClaimItemFeature implements AuthorizationFeature {
```
The showClaimItemFeature is used to show the button for claim an item. The canClaimItemFeature then it is used to verify if the claim can actually be made. This distinction allows first of all to give more feedback to the user if for example he wants to try to make the claim of a profile but this is already associated with another user. Furthermore, it could also allow in the future to adopt more complex logic on the possibility of making a claim.

tdonohue replied:
@LucaGiamminonni : Please add these details/differences to the JavaDocs for these Features then.  Currently, both `CanClaimItemFeature` and `ShowClaimItemFeature` seem extremely similar...and the differences are unclear.  This is the first "Feature" that I'm aware of that has both a `Can*` and a `Show*`.  In other features, usually the UI just checks `Can*` to also determine whether to show the button.  This scenario seems unique, so it needs to be described better.

LucaGiamminonni commented on dspace-api/src/main/java/org/dspace/app/profile/ResearcherProfileServiceImpl.java:
```diff
@@ -0,0 +1,293 @@
+
+    private boolean notClaimableEntityType(final Item item) {
+        String entityType = itemService.getEntityType(item);
+        return !contains(configurationService.getArrayProperty("claimable.entityType"), entityType);
```
I will remove the claimable.entityType to use only the researcher-profile.type property

Commit:
[CST-5306] Improvements on ResearcherProfileRestRepository

Commit:
[CST-5306] Expose researcher-profile.entity-type

LucaGiamminonni commented:
Hi @tdonohue, thanks for the feedbacks. I tried to modify the code with the suggestions you gave me. The main changes I made were:

- moved all authority configuration in the new authority.cfg file
- created a new configuration researcher-profile.cfg with all the profile related properties
- renamed ResourceConflictException in ResourceAlreadyExistsException and mapped with 422 status
- improved java docs
- added a new configuration named researcher-profile.set-new-profile-private to establish if the new profile must be visible or not
- removed claimable.entityType to use only researcher-profile.entity-type property
- improved ResearcherProfileAutomaticClaim checking if the profile type is set and if the item type is compliant to it
- added new test cases

Commit:
[CST-5306] Added eperson_claims discovery configuration to find all claimable profiles

LucaGiamminonni commented:
I also added a discovery configuration (eperson_claims) that will be used on the front-end to suggest any profiles to claim rather than creating a profile from scratch.

Commit:
[CST-5306] Fixed eperson_claims discovery configuration

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResearcherProfileRestRepositoryIT.java:
```diff
@@ -0,0 +1,1093 @@
+        String id = user.getID().toString();
+        String name = user.getName();
+
+        configurationService.setProperty("researcher-profile.collection.uuid", null);
```
For some reason this line seems odd to me.  Why does creation as an Admin require removing the `collection.uuid`?   

I do think we need to have a test that proves what happens when `collection.uuid` is unset (as these tests all set it by default). But, that doesn't appear to be what this test is doing.

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResearcherProfileRestRepositoryIT.java:
```diff
@@ -0,0 +1,1093 @@
+     * @throws Exception
+     */
+    @Test
+    public void testPatchToChangeVisibleAttribute() throws Exception {
```
I might be missing it, but I don't see a test that verifies the behavior of `visible`.  I.e. a test proving that setting it to `true` means an anonymous user can access while setting to `false` means anonymous cannot access (but the user themselves still can & so can admins)

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/ShowClaimItemFeatureIT.java:
```diff
@@ -0,0 +1,177 @@
+                                     .param("feature", showClaimProfileFeature.getName()))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$._embedded").exists())
+                        .andExpect(jsonPath("$.page.totalElements", greaterThanOrEqualTo(1)));
```
Why do we want to show the "claim" button if the Item is already claimed?  Maybe this is what I don't understand about why we need both a ShowClaim and a CanClaim feature.  

(This test appears to be the only difference between ShowClaim & CanClaim...that's why this question is important.)

tdonohue replied:
@LucaGiamminonni : following up on this. I think it's very odd behavior to show a "Claim" button if you cannot actually claim something.  This is why I find this difference between "ShowClaim" and "CanClaim" so odd.  It seems like these two Features should really be checking for the same things (whether you are able to claim the Item). 

If you can claim the item, then the button should appear.  If you cannot, it should not appear.  

It doesn't make sense to be why you'd want the button to appear when an Item is already claimed by someone else (which is what "ShowClaim" currently allows as proven by this test).  As clicking the button will just result in an error from "CanClaim".

I simply don't understand the use case here & why the code is implemented in this manner.

tdonohue reviewed (changes requested):
@LucaGiamminonni : Thanks for your recent updates.  I gave this a second code review and came across a few things you forgot to fix, or need a little more clarification.  I don't think any of these are major though.

I also wanted to note that this PR has  automated feedback from LGTM that needs resolving. Please remember to check your PRs for LGTM feedback and ensure no *new alerts* are listed. This PR currently has two new alerts (one of which though I think is solveable via one of my comments below): https://github.com/DSpace/DSpace/pull/8232#issuecomment-1106594352

Commit:
[CST-5306] Improvements

tdonohue commented on dspace-api/src/main/java/org/dspace/app/profile/ResearcherProfileServiceImpl.java:
```diff
@@ -0,0 +1,322 @@
+
+        if (indexableObjects.size() > 1) {
+            log.warn("Multiple " + profileType + " type collections were found during profile creation");
+            return empty();
```
@LucaGiamminonni : I think we need to rethink this.  In testing this today, I immediately hit strange behavior because of this undocumented requirement.  

If I run this PR with the frontend PR, I get the "Create New" button for a Profile, but if I click it, I get a very confusing HTTP error:
![profile](https://user-images.githubusercontent.com/483997/166809904-497c5b77-98bb-4115-a4a9-fa3af0751645.png)

The cause of the error is that I seem to have multiple Collections configured accepting `Person` entities.  This default setup is going to happen **frequently**, so having Researcher Profiles throw an error by default is not a good behavior.

I'd recommend we either find a way to remove the error that requires sites to have a single Person Collection to create new Researcher Profiles .. or, alternatively, we **require `researcher-profile.collection.uuid` to be set** in order to enable Researcher Profiles.  If we choose to require it, I think we should either throw a more useful error message in the UI, or remove the `Researcher Profile` section if it's not configured.

(Obviously our Submission process already supports the ability to create Person entities in multiple Collections. So, it's odd that Researcher Profiles (which are just Person Entities) needs to behave differently)

tdonohue commented on dspace-api/src/main/java/org/dspace/content/authority/EPersonAuthority.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Override
+    public Choices getMatches(String text, int start, int limit, String locale) {
```
Again, this entire class seems to be missing tests. This is the third time I've pointed this out.  Is there a way to add tests for this class??

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/security/EPersonRestAuthenticationProvider.java:
```diff
@@ -122,6 +135,11 @@ private Authentication authenticateNewLogin(Authentication authentication) {
+
+                        for (PostLoggedInAction action : postLoggedInActions) {
+                            action.loggedIn(newContext);
+                        }
```
In my testing, I ran into a severe bug regarding the `PostLoggedInAction`.  If any of these calls to `loggedIn()` throw an error, it blocks all authentication into the system. 

In my situation, I had forgotten to import the new "dspace.object.owner" field.  But that meant that no one could authenticate into the system.

Here's the stacktrace I saw in the logs:
```
ERROR unknown unknown org.springframework.boot.web.servlet.support.ErrorPageFilter @ Forwarding to error page from
request [/api/authn/login] due to exception [No such metadata field: schema=dspace, element=object, qualifier=owner]
java.lang.IllegalArgumentException: No such metadata field: schema=dspace, element=object, qualifier=owner
        at org.dspace.content.ItemServiceImpl.findByAuthorityValue(ItemServiceImpl.java:1315) ~[dspace-api-7.3-SNAPSHOT.jar:7.3-SNAPSHOT]
        at org.dspace.app.profile.ResearcherProfileServiceImpl.findResearcherProfileItemById(ResearcherProfileServiceImpl.java:223) ~[dspac
e-api-7.3-SNAPSHOT.jar:7.3-SNAPSHOT]
        at org.dspace.app.profile.ResearcherProfileServiceImpl.findById(ResearcherProfileServiceImpl.java:93) ~[dspace-api-7.3-SNAPSHOT.jar
:7.3-SNAPSHOT]
        at org.dspace.app.rest.login.impl.ResearcherProfileAutomaticClaim.currentUserHasAlreadyResearcherProfile(ResearcherProfileAutomatic
Claim.java:112) ~[classes/:7.3-SNAPSHOT]
        at org.dspace.app.rest.login.impl.ResearcherProfileAutomaticClaim.claimProfile(ResearcherProfileAutomaticClaim.java:99) ~[classes/:
7.3-SNAPSHOT]
        at org.dspace.app.rest.login.impl.ResearcherProfileAutomaticClaim.loggedIn(ResearcherProfileAutomaticClaim.java:87) ~[classes/:7.3-
SNAPSHOT]
        at org.dspace.app.rest.security.EPersonRestAuthenticationProvider.authenticateNewLogin(EPersonRestAuthenticationProvider.java:140)
~[classes/:7.3-SNAPSHOT]
        at org.dspace.app.rest.security.EPersonRestAuthenticationProvider.authenticate(EPersonRestAuthenticationProvider.java:89) ~[classes
/:7.3-SNAPSHOT]
```

Would there be some way to catch & log these errors **without** blocking authentication? Perhaps this needs to be in some sort of try/catch block?

tdonohue reviewed (changes requested):
@LucaGiamminonni : I tested this with the frontend PR and it's working overall.  I was able to create a new profile (once I configured a collection, which was required for me -- more on that below), make it public, then make it hidden, then delete it.  All of that worked.

However, I still do have some suggestions below for improvements.  Overall though this backend is looking good.

Commit:
[CST-5306] Removed ShowClaimItemFeature, added EPersonAuthorityIT tests, improved researcher profile's collection search

tdonohue reviewed (approved):
👍 Thank you @LucaGiamminonni !  All my prior feedback has been addressed.  I tested this again today with the frontend PR & the backend appears to be working properly (I did hit a few minor bugs though which I believe are all in the frontend code. See my review on that PR.)

I even tested disabling the `researcher-profile.collection.uuid` while having multiple collections which accept Person entities.  It no longer throws an error. Instead, one of those Collections is just selected and the new Researcher Profile is created there.  In the future we could consider prompting users to select a Collection, but that's not necessary in this initial PR.

tdonohue commented on dspace/config/modules/researcher-profile.cfg:
```diff
@@ -0,0 +1,17 @@
+#---------------------------------------------------------------#
+
+# The Entity Type to use for the Researcher Profile. Defaults to "Person" as this is the recommended Entity Type to use
+researcher-profile.entity-type = Person
```
@LucaGiamminonni : After submitting my review, I just realized one minor thing we may need to do on the backend.  I think we will need to **DISABLE Researcher Profiles by default in DSpace** (which requires commenting this configuration out), simply because Configurable Entities are also disabled by default. 

Another option would be to add a small test to the backend that proves nothing breaks when the configured `entity-type` **doesn't exist**.   In other words, by default, in a DSpace 7 site, the `Person` entity won't exist, so we need to ensure nothing errors out with that setup.  We could prove this via an IT by simply setting `researcher-profile.entity-type = FakeEntity` and ensuring that nothing errors out (as `FakeEntity` won't exist).

Commit:
[CST-5306] Disable researcher profile by default

Commit:
Merge remote-tracking branch '4science-bitbucket/main' into CST-5306

benbosman commented on dspace-api/src/main/java/org/dspace/app/profile/ResearcherProfile.java:
```diff
@@ -0,0 +1,87 @@
+        return UUIDUtils.fromString(dspaceObjectOwner.getAuthority());
+    }
+
+    public String getFullName() {
```
Copying feedback from https://github.com/DSpace/DSpace/pull/8263

Best to mention in the JavaDocs that this name can be outdated, and the latest value should be retrieved from the EPerson instead. This method can best be made deprecated, or removed

benbosman replied:
I think this change wasn't committed yet

benbosman commented on dspace-api/src/main/java/org/dspace/app/profile/ResearcherProfileServiceImpl.java:
```diff
@@ -0,0 +1,317 @@
+    private Optional<Collection> findProfileCollection(Context context) throws SQLException, SearchServiceException {
+        UUID uuid = UUIDUtils.fromString(configurationService.getProperty("researcher-profile.collection.uuid"));
+        if (uuid != null) {
+            return ofNullable(collectionService.find(context, uuid));
```
Is it checked anywhere that this collection has an entity type matching the configured value?

benbosman commented on dspace-api/src/main/java/org/dspace/content/authority/EPersonAuthority.java:
```diff
@@ -0,0 +1,97 @@
+        context = new Context();
+        List<EPerson> ePersons = null;
+        try {
+            ePersons = ePersonService.search(context, text, start, limit);
```
How will this endpoint be made available? I hope it's not possible for submitters to get a list of all EPeople with such a customization

benbosman replied:
Thanks, this limits access already
If I understand it correctly, it does allow every community/collection admin to access it if the property is enabled (not just the community/collection admin of the person collection. But I don't think that's a big issue because community/collection admins can also access the eperson accounts when managing the submitters

@tdonohue do you think we need some warning about that, or can we leave it with this change without warnings?

tdonohue replied:
I think the proposed permission settings are reasonable. As you said, EPersons can already been seen by Community/Collection Admins. So, this aligns with that prior decision.

benbosman reviewed (changes requested):
I've copied my feedback from https://github.com/DSpace/DSpace/pull/8263 which was not handled yet (unless I missed something)
Will review/test the remainder next

benbosman reviewed (changes requested):
I've finished reviewing the code, and added the remainder of my feedback.
I will test next

Commit:
Merge remote-tracking branch '4science-bitbucket/main' into CST-5306

Commit:
[CST-5306] Improved ResearcherProfileServiceImpl and ResearcherProfileRestRepositoryIT tests

Commit:
[CST-5306] Improved CanClaimItemFeatureIT tests

Commit:
[CST-5306] Added security check in EPersonAuthority

Commit:
[CST-5306] Removed getFullName from ResearcherProfile

benbosman reviewed (changes requested):
One final bit of feedback on this PR from the testing

It seems at the moment that I can claim ANY Person account as mine, while I'm a regular user (not admin). It doesn't seem to check on the matching email address

tdonohue commented:
@LucaGiamminonni : After looking back at the code, it appears @benbosman  is correct in [his last comment](https://github.com/DSpace/DSpace/pull/8232#pullrequestreview-975263995).  While the `ResearcherProfileAutomaticClaim` only matches you up if your email matches with the Profile, it looks like it's possible to claim **any profile** manually.  That does seem unexpected, as it'd mean that a user could claim any unclaimed profile they want.   

We likely need to add a test to verify this cannot happen.  My expectation is that a normal user should only be able to claim (either manually or automatically) a profile which matches their email address.  Otherwise, we'd accidentally give normal users to ability to edit any unclaimed profile...they just need to quickly claim it before anyone else.

Beyond that issue, it sounds like the rest of this PR is looking good.  So, if we can find a way to patch up that issue, hopefully this will be ready to merge very soon.

Commit:
[CST-5306] Allow claim only if email matches

tdonohue commented on dspace-api/src/main/java/org/dspace/app/profile/ResearcherProfileServiceImpl.java:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    @Override
+    public ResearcherProfile claim(Context context, EPerson ePerson, URI uri)
```
@LucaGiamminonni : Thanks for updating the CanClaimItemFeature to check for a matching email address.  However, it looks like you forgot to make similar updates to the `claim()` code itself.  So, if you use the REST API directly, it looks like any user can still claim any Profile.

Could you update the REST API claim code, and add ITs to verify that it's not possible for a normal user to claim a profile which doesn't match their email address?

Thanks!

Commit:
[CST-5306] Added email check on ResearcherProfileService claim

benbosman commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResearcherProfileRestRepositoryIT.java:
```diff
@@ -0,0 +1,1135 @@
+    }
+
+    @Test
+    public void testNoAutomaticProfileClaimOccursIfManyClaimableItemsAreFound() throws Exception {
```
@LucaGiamminonni I see you've adjusted this in the code and the tests, but I don't see a test that proves you can't claim an item which doesn't have an email (only a test if the email is different)
Can you add that test (or let me know where it is)?

benbosman commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResearcherProfileRestRepositoryIT.java:
```diff
@@ -0,0 +1,1135 @@
+    }
+
+    @Test
+    public void testNoAutomaticProfileClaimOccursIfTheUserHasAlreadyAProfile() throws Exception {
```
@LucaGiamminonni For this one, a test that it won't be supported if the item has no email should also be added

benbosman commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ResearcherProfileRestRepositoryIT.java:
```diff
@@ -0,0 +1,1135 @@
+
+        String authToken = getAuthToken(user.getEmail(), password);
+
+        getClient(authToken).perform(post("/api/eperson/profiles/")
```
@LucaGiamminonni it seems this is still not adjusted to require the email address to match

benbosman reviewed (changes requested):
Thanks for the updates @LucaGiamminonni 

I have 3 more open comments:
* https://github.com/DSpace/DSpace/pull/8232#discussion_r877864454
* https://github.com/DSpace/DSpace/pull/8232#discussion_r877867940
* https://github.com/DSpace/DSpace/pull/8232#discussion_r877870145

If you can address those, I can re-test

tdonohue commented:
@LucaGiamminonni : When you get the chance, could you please get this updated with @benbosman 's most recent request... it looks like this is very close but has just 3 minor test updates/fixes.  Also, it needs rebasing as it has merge conflicts.

As mentioned last week, I wanted to try and get Researcher Profiles merged this week... but it looks like both this PR and the frontend PR have been accidentally forgotten.  If you can get both updated again, I'd like to re-test as soon as possible so that we can get this work merged.

Commit:
[CST-5306] Added automatic claim test

Commit:
Merge remote-tracking branch '4science-bitbucket/main' into CST-5306

benbosman reviewed (changes requested):
@LucaGiamminonni what I see now is:
* A test that only verifies it's not assigned automatically when there are no two items with the same name. I didn't see a test that ensures it's not assigned automatically when there is only one item
* And for https://github.com/DSpace/DSpace/pull/8232#discussion_r882416552, you have a test that proves it's forbidden to claim if the email address differs, but I don't see any test that proves it's forbidden to claim if the email address is not present on the item

So based on the tests, I might still be able to register in DSpace, claim any person item as mine, adjust e.g. the visibility, or possibly even delete that person item depending on that setting. And I can do this until I've adjusted/deleted all items in the repository that don't have an email field. Although this may not be really possible (I didn't re-test yet), I don't see any prove it's forbidden in the tests, and I know it was possible in the original implementation

Commit:
[CST-5306] Improved claim tests

benbosman commented:
Thanks @LucaGiamminonni , the tests here are really important since this was initially implemented with this use case as an actual bug
I did check `testNoAutomaticProfileClaimOccursIfItemHasNotAnEmail` and it looks good. The last test is also good now

I will run one final manual test next, but it looks promising

benbosman reviewed (approved):
Thanks @LucaGiamminonni 
I've tested the PR as well now, and it looks good to me

tdonohue reviewed (changes requested):
@LucaGiamminonni : Overall this now works GREAT!  However, I found two minor bugs.  

### Regular User Profile is not editable 
If I create a Profile as a regular user, I don't have the permissions to Edit it.  I'm able to successfully view, delete or make it publicly visible.  But, I don't seem to be given WRITE permissions by default.   

If I login as an Admin and look at the created Person Entity, I see that the user was only given READ permissions.  Is this expected behavior?  Or maybe this will be fixed in later PRs?
![read-only](https://user-images.githubusercontent.com/483997/171490881-21f1cddf-63b2-454f-b68c-1f62367233eb.png)

My expectation would be that an individual should be allowed to edit their own profile. (Especially since they can control its visibility and also delete it)

### Profile is missing "person.email" field
A newly created Profile is missing the `person.email` field by default.  While this isn't a major bug, it causes odd behavior if you create & then unlink a profile.
1. Login as a user without a profile
2. Create a new profile
3. Unlink that profile (make sure `hard-delete` is set to false)
4. Now, try to "claim" that same profile.  You cannot, as the profile has no `person.email`, so you can never match with it again.

My expectation is that any profile created by a user should set the `person.email` field by default, especially if that field is required to match the profile for claiming. 

**UPDATE:** It was pointed out in the meeting on June 2 that we likely should also add `person.email` to the hidden fields. So, we should add this to `dspace.cfg` alongside the other hidden fields:
```
metadata.hide.person.email = true
```

Beyond that, I'm OK with the rest of this PR & all my other tests were successful.  I have one other minor suggestion inline below.

Commit:
[CST-5306] Added email and write policy on created researcher profile

tdonohue reviewed (approved):
👍 Thanks @LucaGiamminonni !  This now all works well for me.  I've retested this morning and verified that  `person.email` is now set properly, but it's also hidden to non-Admins (this does have a small side effect that only Admins can edit the email, but that seems "good enough" for an initial implementation).  I've also verified that non-Admins who create a profile do have WRITE access now.  Overall it now works well.

That said, I did find one more small bug...but we could fix this in follow-up work (unless you'd rather fix quickly in this PR):
* When a non-Admin creates a Profile, while they have WRITE access on the Item (giving them the ability to add/edit metadata), they do NOT have WRITE on the ORIGINAL Bundler, which means they currently **cannot upload a bitstream image or change the uploaded image**. But, that could be fixed later.

Overall, this works and looks good. So, I'll go ahead and approve it.  If you want to fix the "WRITE" access to the ORIGINAL bundle, I can review again quickly...otherwise we could fix that in the next PR as needed.

tdonohue commented:
@LucaGiamminonni : This has a merge conflict now (after I merged the Admin Control Panel & Sherpa features today)... could you get it updated with the latest `main` please?  Thanks!

atarix83 committed:
Merge branch 'main' into CST-5306

tdonohue commented:
Merging with +2 approvals
